Improved scripts to create initial conditions files

einola opened the pull request:
# Improved scripts to create initial conditions files

Fixes #945

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

Adds the functionality to make_init_base.py to make generic geographic grids, in addition to idealised ones. The user must supply lat/lon coordinates of the grid and may provide a mask and initial fields. The program can create nextsim grids from a grid of lat/lon coordinates or lat/lon coordinates of a NEMO C-grid (which also provides coordinates of the upper right corner of the grid box).

    Usage:
     0. Import make_init_base
      >> from make_init_base import initMaker
     1. Create an initialiser object with a given filename, e.g.
      >> init = initMaker("test")
         2a. Create a cartesian grid from dimensions and resolution, e.g.
          >> init.make_cartesian_grid(128, 128, 3e3)
         2b. Create a geographic grid from a file containing coordinates. We need to define type of grid:
            * "p": only centre points of the grid are given
            * "ur": the centre and upper right corner of the grid is given
            * "ll": the centre and lower left corner of the grid is given
          init.make_geographic_grid("coordinates.nc", "ur", plat_name="gphit", plon_name="glamt", qlat_name="gphif", qlon_name="glamf")
     3. Modify any variables needed, e.g.
      >> init.cice = 1
      >> init.hice = 3
      >> init.hice[10:20, 10:20] = 4
     4. The init file is written when the initialiser object goes out of scope (e.g. when the program ends or at the end of a loop).

I also modified and renamed make_initNH.py to use make_init_base.py

---
# Test Description

I used the script to recreate the grid of the integration test, as well as the grid from our NEMO-coupled setup and a few others. All of which run successfully. 

---
# Documentation Impact

The functionality of make_init_base.py should be documented in the usage section of the manual, but it is not.

---
# Other Details

N/A

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] This change conforms to the conventions described in the README

einola committed:
Make space for geographic grids in make_init_base.py

Just rewrite it a bit so that you have to call either
make_cartesian_grid or make_geographic_grid before setting your
variables. Only make_cartesian_grid implemented so far.

einola committed:
Read-in support for NEMO files

make_init_base.py can now handle NEMO files and produce vertex
coordinates and the grid azimuth angle on a displaced pole.

einola committed:
Clean make_init_base.py

Doc strings and miscellaneous cleaning. I also added two functions to
get the grid centre longitude and latitude.

einola committed:
Rewrite make_initNH.py to use make_init_base.py

The script is now much shorter and also more generalised. But it expects
 the grid file to follow NEMO naming conventions for the coordinates.

einola committed:
Fix some comments

This belongs with the commit before last.

einola committed:
Code to handle initialising from a regular grid

make_init_base.py can now handle grids that only have lat/lon
information at the grid cell centre. I do this by assuming everything is
regular and that the vertices are +/- dx and dy away from the grid cell
centre. I also updated make_init_geographic.py to be able to handle such
grids.

einola committed:
Fixing a bit comments and signs

The grid is up-side-down and we also need the reverse of the angle. Two
minuses that make a plus, but now it's clearer why it all works (I
hope).

einola committed:
Add a option for double closed boundaries in make_init_geographic.py

I think it's useful for our NANUK setup.

einola committed:
Using netCDF missing value for the mask of the init file

The velocities need to have zeros on land, but the rest can have missing
values.

einola committed:
Merge branch 'develop' into issue945_init_scripts

# Conflicts:
#	run/interpolators.py
#	run/make_initNH.py
#	run/make_init_base.py

einola committed:
Run ruff on modified files

I did "ruff format --config ruff.toml ./" in the project base directory.
This modified a lot of files, but I just picked the ones relevant for
this PR.

einola committed:
More ruff fixes

This time, I did "ruff check --diff ./ > patch.diff" in the run
directory, followed by "git apply patch.diff". That gave a different
result to "ruff format".
.

einola commented:
~I got into a fight with Ruff, and this resulted in a lot of unrelated formatting changes - sorry.~

I won over Ruff in the end!

einola committed:
Revert "More ruff fixes"

This reverts commit 6ec669442924f9017dfe510543c3864d1011be3d.

einola committed:
Revert "Run ruff on modified files"

This reverts commit 9f8f9f154eb469a4d87d8476f7bf6ef5e67ea00d.

einola committed:
Trying to make ruff happy

by doing "ruff check --diff ./ > patch.diff" in the run directory and
then "git apply patch.diff".

joewallwork commented on run/make_init_base.py:
```diff
@@ -63,140 +46,459 @@ def __init__(self, fname, nFirst, nSecond, res, nCg=1, nDg=1, nDgStress=3, nCoor
+                print("Warning: '" + check[0] + "' is all zeros (this may be ok, if that's what you want).")
+
+            if not check[2]:
+                print("Error: '" + check[0] + "' is the wrong shape")
```
This would be more readable with
```suggestion
        for (field, allZero, wrongShape) in [["cice", (self.cice == 0).all(), self.cice.shape == (self.__nFirst, self.__nSecond)],
                      ["hice", (self.hice == 0).all(), self.hice.shape == (self.__nFirst, self.__nSecond)],
                      ["hsnow", (self.hsnow == 0).all(), self.hsnow.shape == (self.__nFirst, self.__nSecond)],
                      ["sss", (self.sss == 0).all(), self.sss.shape == (self.__nFirst, self.__nSecond)],
                      ["sst", (self.sst == 0).all(), self.sst.shape == (self.__nFirst, self.__nSecond)]]:

            if self.__checkZeros and allZero:
                print(f"Warning: '{field}' is all zeros (this may be ok, if that's what you want).")

            if not wrongShape:
                print(f"Error: '{field}' is the wrong shape")
```

joewallwork commented on run/make_init_geographic.py:
```diff
@@ -0,0 +1,98 @@
+from interpolators import topaz4_interpolate
+from make_init_base import initMaker
+
+topaz_mdi = -32767
```
Presumably this is correct?
```suggestion
# Missing data value for TOPAZ data format
topaz_mdi = -32767
```

einola replied:
Yes, technically, but ``topaz_mdi`` isn't used by this script - it's a leftover from ``make_initNH.py``. So, I've deleted this line now.

joewallwork commented on run/make_init_geographic.py:
```diff
@@ -0,0 +1,98 @@
+    parser.add_argument("--grid-file", dest="grid_file", default="25km_NH.nc", help="Path of the input grid file.")
+    parser.add_argument("--mask-file", dest="mask_file", default="25km_NH.nc", help="Path of the input mask file.")
+    parser.add_argument("--topaz-path", dest = "topaz_path", default=".", help = "Path containing the TOPAZ4 files.")
+    parser.add_argument("--boundary", dest = "boundary", default="open", help='One of "open" (default), "closed", or "double".')
```
Out of interest, what does 'double' boundary mean?

einola replied:
It's a closed boundary with two rows or columns of land at the boundary. For backwards compatibility with some obscure setups ;)

einola replied:
It feels a bit excessive to explain this in the terse online help, so I'm leaving this unchanged.

joewallwork reviewed (approved):
This looks good to me, all of my comments are minor and optional.

One is that the "dunder" double-underscore methods (e.g., `__init__`) in Python are only usually used for customising built-in functionality (see https://www.pythonmorsels.com/every-dunder-method/), whereas here it's being used for custom attributes and methods. If the intention is to mark those attributes or methods as "private" (insofar as Python has any concept of privacy), the usual convention is to start the name with a single underscore. This isn't a big deal but I just thought I'd point it out. Also, I realise that this notation was already being used before this PR.

einola committed:
Remove double underscores

The "private" methods and variables were marked by double underscores,
but this is not the convention (see Joe's comments on PR #946)

einola committed:
Improve the readability of make_init_base.py

Loop through a tuple when checking fields, so that you don't have to guess what check[0] and check[1] stands for.

Co-authored-by: Joe Wallwork <22053413+jwallwork23@users.noreply.github.com>

einola committed:
Remove the topaz_mdi variable

It's not used any more.

einola committed:
Ran ruff format relevant python scripts

Joe used ruff to reformat all python files, resulting in a conflict with
this branch. I'm hoping that running ruff format on my changed files
and then merging with develop will be painless.

einola committed:
Merge remote-tracking branch 'origin/develop' into issue945_init_scripts

einola committed:
More ruff format

I seem to have missed something when formatting earlier. I now ran ruff
format run/*.py, so _everything_ should be fine.
